packages/rum-recorder/src/domain/segmentCollection.ts

```diff
@@ -0,0 +1,161 @@
+  send: (data: Uint8Array, meta: SegmentMeta) => void
+) {
+  const worker = createDeflateWorker()
+  return doStartSegmentCollection(
+    lifeCycle,
+    () => doGetSegmentContext(applicationId, session, parentContexts),
+    send,
+    worker
+  )
+}
+
+export function doStartSegmentCollection(
+  lifeCycle: LifeCycle,
+  getSegmentContext: () => SegmentContext | undefined,
+  send: (data: Uint8Array, meta: SegmentMeta) => void,
+  worker: DeflateWorker,
+  emitter: EventEmitter = window
+) {
+  let currentSegment: Segment | undefined
+  let currentSegmentExpirationTimeoutId: ReturnType<typeof setTimeout>
+  let nextSegmentCreationReason: CreationReason = 'init'
+
+  const writer = new DeflateSegmentWriter(
+    worker,
+    (size) => {
+      if (size > SEND_BEACON_BYTE_LENGTH_LIMIT) {
+        flushSegment('max_size')
+      }
+    },
+    (data, meta) => {
+      send(data, meta)
+    }
+  )
+
+  // Flush when the RUM view changes
+  const { unsubscribe: unsubscribeViewCreated } = lifeCycle.subscribe(LifeCycleEventType.VIEW_CREATED, () => {
+    flushSegment('view_change')
+  })
+
+  // Flush when the session is renewed
+  const { unsubscribe: unsubscribeSessionRenewed } = lifeCycle.subscribe(LifeCycleEventType.SESSION_RENEWED, () => {
+    flushSegment('session_renewed')
+  })
+
+  // Flush when leaving the page
+  const { unsubscribe: unsubscribeBeforeUnload } = lifeCycle.subscribe(LifeCycleEventType.BEFORE_UNLOAD, () => {
+    flushSegment('before_unload')
+  })
+
+  // Flush when visibility changes
+  const { stop: unsubscribeVisibilityChange } = addEventListener(
+    emitter,
+    DOM_EVENT.VISIBILITY_CHANGE,
+    () => {
+      if (document.visibilityState === 'hidden') {
+        flushSegment('visibility_change')
+      }
+    },
+    { capture: true }
+  )
+
+  function flushSegment(creationReason: CreationReason) {
+    if (currentSegment) {
+      currentSegment.flush()
+      currentSegment = undefined
+      clearTimeout(currentSegmentExpirationTimeoutId)
+    }
+
+    nextSegmentCreationReason = creationReason
+  }
+
+  return {
+    addRecord(record: Record) {
+      if (!currentSegment) {
+        const context = getSegmentContext()
+        if (!context) {
+          return
+        }
+
+        currentSegment = new Segment(writer, context, nextSegmentCreationReason, record)
+        // Replace the newly created segment after MAX_SEGMENT_DURATION
+        currentSegmentExpirationTimeoutId = setTimeout(
+          monitor(() => {
+            flushSegment('max_duration')
+          }),
+          MAX_SEGMENT_DURATION
+        )
+      } else {
+        currentSegment.addRecord(record)
+      }
+    },
+    stop() {
+      unsubscribeViewCreated()
+      unsubscribeBeforeUnload()
+      unsubscribeVisibilityChange()
+      unsubscribeSessionRenewed()
+      worker.terminate()
+    },
+  }
+}
+
+export function doGetSegmentContext(applicationId: string, session: RumSession, parentContexts: ParentContexts) {
```

Comment:
```suggestion
export function computeSegmentContext(applicationId: string, session: RumSession, parentContexts: ParentContexts) {
```
"do get" seems a bit weird to me

Reply:
All "do*" functions seems weird to me to be honest!

👍 I'll change it to `compute`